### PR TITLE
policy: Ephemeral anchors

### PR DESCRIPTION
**Pull request by instagibbs**
TODO:

I'll be updating to segwit variant soon, with updated BIP text.

-----

Builds on top of https://github.com/bitcoin/bitcoin/pull/25038 for consideration of inclusion to the proposal. Requires V3, for simplicity of reasoning and usage. Implementation of idea written out at https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2022-October/021036.html

BIP text here: https://github.com/instagibbs/bips/blob/ephemeral_anchor/bip-ephemeralanchors.mediawiki

### CODE REVIEW

**Comment by luke-jr** [comment]
Concept NACK as a mere policy. Without consensus enforcement, cost is a full UTXO which may potentially never be spent. `OP_TRUE` does not have a weight that reflects this high cost.

**Comment by instagibbs** [comment]
@luke-jr Can you elaborate more on what the potential issues are?

This policy defines that any transaction with such an output must be 0-fee, which means there's little incentive to mine it on its own, but only as part of a CPFP spend of that output.

Assuming the PR is correct and being run on the network, there should be no way(aside from miners prioritizing the parent explicitly) to leaving the utxo in the uxto set. Prioritizing the parent could also be disallowed easily.

**Comment by luke-jr** [comment]
The main issue I see is that it has a very low weight, but rather high cost. There is no consensus rule requiring it to be spent immediately.

Couldn't it be CPFP'd by spending another of its outputs?

**Comment by instagibbs** [comment]
The argument here is that dust limits are already policy only, and this is basically a "cut-through" of  potentially dust outputs. It's no practical change of dust policy, except in the case where a miner prioritizes a parent tx with this kind of output after its included in mempool(it's in the test cases)

**Comment by JeremyRubin** [comment]
@luke-jr  either you do a consensus fork and everything that entails for something like tx sponsors, or your enable in policy only something like ephemeral anchors.

One option that could be taken would be to make the OP_TRUE script contain an upgradable NOP that we document as reserved to in the future treat as a pre-execution parsed flag with a consensus must-spend rule, and then we'd be able to (if there was wide adoption of ephemeral anchors and we saw utxo bloat) add a consensus rule to enforce them as must-spend.

**Comment by luke-jr** [comment]
>One option that could be taken would be to make the OP_TRUE script contain an upgradable NOP that we document as reserved to in the future treat as a pre-execution parsed flag with a consensus must-spend rule, and then we'd be able to (if there was wide adoption of ephemeral anchors and we saw utxo bloat) add a consensus rule to enforce them as must-spend.

I think this would be better, since it allows a real OP_TRUE to be given a more cost-based weight in a future softfork.

**Comment by MarcoFalke** [comment]
It would also be possible to deploy a consensus change to treat them as unspendable, unless they are spent in the block they are created in, but it seems odd to enforce policy with consensus, and this is likely the wrong thread to discuss anyway.

**Comment by JeremyRubin** [comment]
@MarcoFalke that is an inverse timelock, which has nasty consequences for reorg safety and dependent transactions.

**Comment by instagibbs** [comment]
@JeremyRubin including a NOP would just be to not "camp" a raw single byte value push for a future soft fork? Making sure I didn't miss the underlying motivation.

**Comment by JeremyRubin** [comment]
more or less, you might also be able to do a 0 byte segwit v1 script or something too.

Otherwise you end up with some "specific script template matching upgrade", which is IMO messy & risks being uneccessarily confiscatory.

**Comment by ajtowns** [comment]
Camping an OP_NOP and preventing it from being used for any other future soft fork seems worse than camping OP_TRUE.

If in the future we have a soft fork that makes small scriptPubKeys more costly to limit utxo bloat with an exemption for ephmeral outputs that are spent in the same block, which should just apply both rules to all scriptPubKeys, whether they're OP_TRUE, OP_2, or OP_NOP5 OP_TRUE...

**Comment by instagibbs** [comment]
I'll rebase on top of https://github.com/bitcoin/bitcoin/pull/25038 once it leaves draft mode, since I have to update my tests to new V3 packages.

**Comment by naumenkogs** [comment]
What will happen at the p2p level, if the attacker manages to submit a transaction with ephemeral outputs to many nodes in the network without a child?

It looks like its wtxid will make it into `m_recent_rejects`, and then the network won't reconsider the tx even when the child is attached?

**Comment by instagibbs** [comment]
@naumenkogs Ideally I would hope it would be treated like any other too-low-fee transaction, and be re-evaluated again in the package context. I'll double check, thanks!

File: test/functional/mempool_ephemeral_anchor.py [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+        node.submitpackage(package_hex3)
+        self.assert_mempool_contents(expected=package_txns3, unexpected=[])
+
+        # Same parent coins/txs, don't spend all anchors, with same parent! Should reject RBF.
+        package_hex4, package_txns4 = self.create_simple_package(parent_coins=parent_coins, parent_fee=0, child_fee=DEFAULT_FEE*2, version=3, spend_anchors=2)
+        assert_raises_rpc_error(-26, "v3-tx-nonstandard", node.submitpackage, package_hex4)
+        self.assert_mempool_contents(expected=package_txns3, unexpected=[])
+
+        # All ephemeral anchors this time, make sure old child evicted
+        package_hex5, package_txns5 = self.create_simple_package(parent_coins=parent_coins, parent_fee=0, child_fee=DEFAULT_FEE*2, version=3)
```
Comment (instagibbs) [inline test/functional/mempool_ephemeral_anchor.py:151]: todo: double-check we can't get blinded to a parent tx if submitted separately

File: test/functional/mempool_ephemeral_anchor.py [hunk 2]
```diff
@@ -331,11 +331,12 @@ def test_xor_rbf(self):
         self.generate(node, 1)
         self.assert_mempool_contents(expected=[second_package_txns[1]], unexpected=[])
 
-        # Parent can still be prioritised
+        # Parent can't be prioritised; RPC call succeeds but is modified fees are ignored if it has EAs
```
Comment (ajtowns) [inline test/functional/mempool_ephemeral_anchor.py:334]: "i***t***s modified fees" ?

File: src/txmempool.cpp [hunk 3]
```diff
@@ -402,6 +407,11 @@ void CTxMemPoolEntry::UpdateDescendantState(int64_t modifySize, CAmount modifyFe
 
 void CTxMemPoolEntry::UpdateAncestorState(int64_t modifySize, CAmount modifyFee, int64_t modifyCount, int64_t modifySigOps)
 {
+    // We don't want to encourage miners to mine these transacitons alone. Instead
```
Comment (ajtowns) [inline src/txmempool.cpp:410]: "transac***ti***ons"

Comment (instagibbs) [reply src/txmempool.cpp]: no longer changing this

File: test/functional/mempool_ephemeral_anchor.py (hunk 2 above)
Comment (instagibbs) [reply test/functional/mempool_ephemeral_anchor.py]: removed the test since it's duplicated by v3 work which now trims lower than min relay fee txns

**Comment by instagibbs** [comment]
Rebased on top of latest https://github.com/bitcoin/bitcoin/pull/25038

1) simplified spends checking due to 2-txn topology for better error reporting
2) simplified tests, no longer messing with test utils at all

**Comment by instagibbs** [comment]
alternative commit that makes it OP_TRUE instead: https://github.com/instagibbs/bitcoin/commit/bf7676da978188ebf759288a147f79eb41bf80cc

**Comment by instagibbs** [comment]
Switched to OP_TRUE, as it was less confusing for at least a few people, and only requires light test editing. Had to rebase on top of master for trivial merge conflict.

File: src/policy/policy.cpp [hunk 4]
```diff
@@ -142,8 +142,12 @@ bool IsStandardTx(const CTransaction& tx, const std::optional<unsigned>& max_dat
             reason = "bare-multisig";
             return false;
         } else if (IsDust(txout, dust_relay_fee)) {
-            reason = "dust";
-            return false;
+            // We ensure ephemeral output is spent in the same mempool package
```
Comment (instagibbs) [inline src/policy/policy.cpp:145]: suggested in other PR: "switch the ordering? else if (whichType != ANCHOR && IsDust())"

**Comment by instagibbs** [comment]
I reworked this PR with changes incorporated from https://github.com/bitcoin-inquisition/bitcoin/pull/23

**Comment by instagibbs** [comment]
Should add a test showing that node restart dumps ephemeral transactions since re-admission to the mempool is done single tx by single tx only.

**Comment by instagibbs** [comment]
https://github.com/bitcoin/bitcoin/pull/26403/commits/a1982f5773690bc0c74d1bb932cb7552b00aacaf shows that current re-submission logic on node restart fails to evict ephemeral transaction. 

Until there is package-aware re-submission, should these transactions just be dropped on the floor? These types of transactions are opting into fairly aggressive fee bumping patterns, maybe that's an ok first cut.

**Comment by instagibbs** [comment]
note to self: after rebased onto BIP331, need to make sure rejected ephemeral anchor tx is put in `m_recent_rejects_reconsiderable` filter instead of `m_recent_rejects`

File: src/validation.cpp [hunk 5]
```diff
@@ -834,6 +834,11 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
         return false; // state filled in by CheckTxInputs
     }
 
+    // We check for 0-base-fee transaction here now that we have access to ws.m_base_fees.
+    if (HasPayToAnchor(tx) && ws.m_base_fees != 0) {
+        return state.Invalid(TxValidationResult::TX_NOT_STANDARD, "invalid-ephemeral-fee");
```
Comment (instagibbs) [inline src/validation.cpp]: needs to return a code like `TX_LOW_FEE` from bip331 post-rebase

File: src/policy/v3_policy.cpp [hunk 6]
```diff
@@ -126,13 +126,19 @@ std::optional<std::string> CheckEphemeralSpends(const CTransactionRef& ptx,
...
 {
     // No anchor; it's ok
     if (!HasPayToAnchor(tx)) {
         return true;
     }
 
+    /* Only allow in package feerate context */
+    if (!package_context) {
+        /* Allows re-evaluation in package context */
+        return state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "missing-ephemeral-spends");
```
Comment (instagibbs) [inline src/policy/v3_policy.cpp:139]: should return code like `TX_LOW_FEE` re:BIP331 to avoid blinding relay of an otherwise valid parent

**Comment by instagibbs** [comment]
Random thought:

OP_TRUE is txid-malleable by block maker. e.g., they could theoretically add in superfluous pushes in the `scriptSig` and this would not fail validation because `CLEANSTACK` is a policy-only rule.

In this particular V3 context, it just means your CPFP could change txid, wouldn't do anything else. Maybe in a more general context if we weren't restricted to V3 topology this would be an issue? In the context of keyless anchors, I'm not sure that will ever make sense anyways.

The "fix" would be: `OP_DEPTH OP_0 OP_EQUAL`, making it consensus-non-malleable at the cost of 8 additional WU. Seems like a bad tradeoff, noting anyways.

**Comment by glozow** [comment]
> @instagibbs pushed 1 commit.
>    [faeed68](https://github.com/bitcoin/bitcoin/commit/faeed687e5cde5e32750d93818dd1d4add837f24) nice crash bro

hacked a quick-ish fix on top of this branch [here](https://github.com/glozow/bitcoin/tree/2023-08-10-fix-crash-bro). As discussed offline, the general problem is with calling `TrimToSize` in the middle of package validation (meaning something can go from already-in-mempool to evicted, or just-submitted to evicted) so a coin may be cached in `m_view` but then it disappears from the mempool/temporary coins in `m_viewmempool`. Crash is in `CheckInputsFromMempoolAndCache` which asserts the coin exists, but it doesn't anymore. Will open a separate PR to address this since it's also a problem pre-v3.

**Comment by whitslack** [comment]
What's the point of having a non-malleable version? Okay, so a miner cannot add extra pushes to the input script that spends the anchor, but there is nothing to stop the miner from mining an alternative child transaction that spends *only* the anchor. Thus, the party that had broadcast the original child transaction will be forced to construct a new spending transaction with a different TxID anyway. Indeed, anyone (not only a miner) is free to propose a replacement child that spends *only* the anchor, so it will never be okay to depend upon the TxID of a transaction that spends an ephemeral anchor. So again I ask, what is the point of having a non-malleable ephemeral anchor?

**Comment by instagibbs** [comment]
> Thus, the party that had broadcast the original child transaction will be forced to construct a new spending transaction with a different TxID anyway

Canonical example here which motivated the change was that of LN channel splicing being used to cpfp bump another transaction, would could be another channel's commitment transaction itself. This requires the ephemeral anchor-spending tx(the splice) to retain txid stability otherwise funds are locked up in the new funding output.

I understand that not all use-cases require such things, but two points:
1) It's easy to expand standardness, and really hard to restrict. In deploying one version, we'll likely learn lessons we'd want to apply to other versions. If we see tons of update from wallets that explicitly don't care about txid stability? That's a signal as well.
2) "I don't see the problem" is much weaker than "there is no problem" and I'd rather be able to reason via modern bitcoin rules

**Comment by whitslack** [comment]
> This requires the ephemeral anchor-spending tx(the splice) to retain txid stability otherwise funds are locked up in the new funding output.

The point I was missing is that a non-malleable anchor ensures that the *other* output(s) do **not** get spent in a transaction whose TxID was not known in advance of signing. Thank you for the answer.

**Pull request closed without merging**